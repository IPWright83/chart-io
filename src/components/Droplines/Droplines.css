.chart-it.dropline {
   stroke-width: 2;
   stroke-dasharray: 4;
   pointer-events: none;
}
