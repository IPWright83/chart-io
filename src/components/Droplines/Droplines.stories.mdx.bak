#<!--- Droplines.stories.mdx --->

#

#import { Meta, Story, Canvas } from "@storybook/addon-docs/blocks";

#

#import { Droplines } from "./index";
#import { Chart } from "../Chart";

#

#<Meta title="Components/Droplines" component={Droplines} />

#

#export const DroplinesTemplate = (args) => (
#<Chart
#onStoreCreated={(store) => {
#store.dispatch({
#type: "EVENT.ADD_DROPLINE",
#payload: {
#isHorizontal: true,
#color: "steelblue",
#x1: 150,
#x2: 0,
#y1: 150,
#y2: 150,
#}
#});

#

            #store.dispatch({
                #type: "EVENT.ADD_DROPLINE",
                #payload: {
                    #isVertical: true,
                    #color: "steelblue",
                    #x1: 150,
                    #x2: 150,
                    #y1: 150,
                    #y2: 0,
                #}
            #});
        #}}
    #>
        #<Droplines />
    #</Chart>

#

#);

#

## Droplines Component

#

#The droplines component is a designed to show visual feedback when someone hovers over a data point within a chart. It allows both horizontal and/or vertical lines to be drawn by updating the internal redux store.

#

#<!-- const onStoreCreated = (store) => {
#// The store can now be acccessed. If you need dynamic access store
#// this value using a React useState hook.
#};

#

#<Chart onStoreCreated={onStoreCreated} ...>
#... #</Chart> -->

#

#<Story name="Droplines" args={{}}>
#{DroplinesTemplate.bind({})} #</Story>

#

#### Props

#

#| Prop | Type | Default | Note |
#| ---------------- | --------- | ------- | --------------------------------------- |
#| `showVertical` | `boolean` | true | Wether vertical lines should be drawn |
#| `showHorizontal` | `boolean` | true | Wether horizontal lines should be drawn |

#

#<br />
