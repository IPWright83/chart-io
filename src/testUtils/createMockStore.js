/**
 * Creates a mock Redux store for testing with
 * @param  {Object} state     The initial state of the store
 * @return {Object}           A mock store
 */
export const createMockStore = (state) => {
    return {
        getState: () => state,
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    };
};
