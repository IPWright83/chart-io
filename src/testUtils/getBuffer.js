// import { toMatchImageSnapshot } from "jest-image-snapshot";

// expect.extend({ toMatchImageSnapshot });

// *
//  * Ensure a canvas matches the given snapshot
//  * @param  {HTMLElement} canvas         The canvas object

export const getBuffer = (canvas) => {
    const image = canvas.toDataURL("image/png");
    const data = image.replace(/^data:image\/\w+;base64,/, "");
    return Buffer.from(data, "base64");
};
