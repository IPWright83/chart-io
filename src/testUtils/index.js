export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
