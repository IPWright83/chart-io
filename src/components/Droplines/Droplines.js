import "./Droplines.css";

import * as d3 from "d3";
import PropTypes from "prop-types";
import { useEffect } from "react";
import { useSelector } from "react-redux";

import { chartSelectors, eventSelectors } from "../../store";

/**
 * This component renders the droplines that are triggered from various plots
 * @return {ReactElement}  The Background component
 */
const Droplines = ({ layer, showVertical = true, showHorizontal = true }) => {
    const animationDuration = useSelector((s) => chartSelectors.animationDuration(s));
    const droplines = useSelector((s) => eventSelectors.droplines(s)).filter(
        (dl) => (dl.isVertical && showVertical) || (dl.isHorizontal && showHorizontal)
    );

    useEffect(() => {
        if (!layer.current) return;

        const join = d3
            .select(layer.current)
            .selectAll(".dropline")
            .data(droplines, (d) => `${d.x1}-${d.x2}-${d.y1}-${d.y2}`);

        join.exit().remove();

        // Add any new droplines
        join.enter()
            .append("line")
            .attr("class", "dropline")
            .style("stroke", (d) => d.color)
            .attr("x1", (d) => d.x1)
            .attr("x2", (d) => d.x1)
            .attr("y1", (d) => d.y1)
            .attr("y2", (d) => d.y1)
            .transition()
            .delay(animationDuration)
            .duration(animationDuration)
            .attr("x2", (d) => d.x2)
            .attr("y2", (d) => d.y2);
    }, [animationDuration, layer, droplines]);

    return null;
};

Droplines.propTypes = {
    /**
     * The layer to be rendered upon. Typically this is an `<svg:g>` or a fake HTMLElement when using canvas.
     * @default undefined
     * @type {HTMLElement}
     */
    layer: PropTypes.object,

    /**
     * Should horizontal droplines be shown?
     * @default true
     * @type {Boolean}
     */
    showHorizontal: PropTypes.bool,

    /**
     * Should vertical droplines be shown?
     * @default true
     * @type {Boolean}
     */
    showVertical: PropTypes.bool,
};

export { Droplines };
