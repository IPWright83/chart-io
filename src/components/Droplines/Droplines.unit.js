import React from "react";
import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import { render } from "@testing-library/react";

import { Droplines } from "./Droplines";

describe("Droplines", () => {
    const store = {
        getState: () => ({
            event: {
                droplines: [
                    { isHorizontal: true, color: "red", x1: 50, x2: 0, y1: 50, y2: 50 },
                    { isVertical: true, color: "red", x1: 50, x2: 50, y1: 50, y2: 0 },
                ],
            },
        }),
        dispatch: () => {},
        subscribe: () => {},
    };

    it("should render correctly", async () => {
        const layer = { current: document.createElement("custom") };

        render(
            <Provider store={store}>
                <Droplines layer={layer} />
            </Provider>
        );

        expect(layer.current).toMatchSnapshot();
    });

    it("should render just vertical lines correctly", () => {
        const layer = { current: document.createElement("custom") };

        TestRenderer.act(() => {
            TestRenderer.create(
                <Provider store={store}>
                    <Droplines layer={layer} showHorizontal={false} />
                </Provider>
            ).toJSON();
        });

        expect(layer.current).toMatchSnapshot();
    });

    it("should render just horizontal lines correctly", () => {
        const layer = { current: document.createElement("custom") };

        TestRenderer.act(() => {
            TestRenderer.create(
                <Provider store={store}>
                    <Droplines layer={layer} showVertical={false} />
                </Provider>
            ).toJSON();
        });

        expect(layer.current).toMatchSnapshot();
    });

    it("should skip if there is no layer avaliable", () => {
        const layer = { current: null };
        TestRenderer.act(() => {
            TestRenderer.create(
                <Provider store={store}>
                    <Droplines layer={layer} />
                </Provider>
            ).toJSON();
        });

        // Should be empty
        expect(layer.current).toMatchSnapshot();
    });
});
