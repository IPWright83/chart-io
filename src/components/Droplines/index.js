import { withSVG } from "../../hoc";
import { Droplines as DroplinesBase } from "./Droplines";

const Droplines = withSVG(DroplinesBase);

export { Droplines };
