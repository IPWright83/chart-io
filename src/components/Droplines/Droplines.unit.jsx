import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import { createMockStore } from "../../testUtils";

import { Droplines } from ".";
import { Droplines as DroplinesBase } from "./Droplines";

describe("Droplines", () => {
    const store = createMockStore({
        event: {
            droplines: [
                { isHorizontal: true, color: "red", x1: 50, x2: 0, y1: 50, y2: 50 },
                { isVertical: true, color: "red", x1: 50, x2: 50, y1: 50, y2: 0 },
            ],
        },
    });

    it("should render correctly", async () => {
        const layer = { current: document.createElement("custom") };

        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <Droplines />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render just vertical lines correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <Droplines showHorizontal={false} />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render just horizontal lines correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <Droplines showVertical={false} />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should skip if there is no layer avaliable", () => {
        const layer = { current: null };

        render(
            <Provider store={store}>
                <svg>
                    <DroplinesBase layer={layer} />
                </svg>
            </Provider>,
        );

        // Should be empty
        expect(layer).toMatchSnapshot();
    });
});
