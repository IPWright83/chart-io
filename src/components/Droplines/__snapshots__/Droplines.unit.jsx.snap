// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Droplines should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-droplines"
    >
      <line
        class="chart-it dropline"
        style="stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
      <line
        class="chart-it dropline"
        style="stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Droplines should render just horizontal lines correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-droplines"
    >
      <line
        class="chart-it dropline"
        style="stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Droplines should render just vertical lines correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-droplines"
    >
      <line
        class="chart-it dropline"
        style="stroke: red;"
        x1="50"
        x2="50"
        y1="50"
        y2="50"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`Droplines should skip if there is no layer avaliable 1`] = `
Object {
  "current": null,
}
`;
